xarray/core/indexes.py

```diff
@@ -1,12 +1,202 @@
+        if isinstance(coord_names, Iterable) and not isinstance(coord_names, str):
+            self.coord_names = tuple(coord_names)
+        else:
+            self.coord_names = tuple([coord_names])
+
+    @classmethod
+    def from_variables(
+        cls, variables: Dict[Hashable, "Variable"], **kwargs
+    ):  # pragma: no cover
+        raise NotImplementedError()
+
+    def to_pandas_index(self) -> pd.Index:
+        """Cast this xarray index to a pandas.Index object or raise a TypeError
+        if this is not supported.
+
+        This method is used by all xarray operations that expect/require a
+        pandas.Index object.
+
+        """
+        raise TypeError(f"{type(self)} cannot be cast to a pandas.Index object.")
+
+    def equals(self, other):  # pragma: no cover
+        raise NotImplementedError()
+
+    def union(self, other):  # pragma: no cover
+        raise NotImplementedError()
+
+    def intersection(self, other):  # pragma: no cover
+        raise NotImplementedError()
+
+
+class PandasIndex(Index, ExplicitlyIndexedNDArrayMixin):
+    """Wrap a pandas.Index to preserve dtypes and handle explicit indexing."""
+
+    __slots__ = ("array", "_dtype")
+
+    def __init__(
+        self, array: Any, dtype: DTypeLike = None, coord_name: Optional[Hashable] = None
+    ):
+        if coord_name is None:
+            coord_name = tuple()
+        super().__init__(coord_name)
+
+        self.array = utils.safe_cast_to_index(array)
+
+        if dtype is None:
+            if isinstance(array, pd.PeriodIndex):
+                dtype_ = np.dtype("O")
+            elif hasattr(array, "categories"):
+                # category isn't a real numpy dtype
+                dtype_ = array.categories.dtype
+            elif not utils.is_valid_numpy_dtype(array.dtype):
+                dtype_ = np.dtype("O")
+            else:
+                dtype_ = array.dtype
+        else:
+            dtype_ = np.dtype(dtype)
+        self._dtype = dtype_
+
+    @classmethod
+    def from_variables(cls, variables: Dict[Hashable, "Variable"], **kwargs):
+        if len(variables) > 1:
+            raise ValueError("Cannot set a pandas.Index from more than one variable")
+
+        varname, var = list(variables.items())[0]
+        return cls(var.data, dtype=var.dtype, coord_name=varname)
+
+    def to_pandas_index(self) -> pd.Index:
+        return self.array
+
+    @property
+    def dtype(self) -> np.dtype:
+        return self._dtype
+
+    def __array__(self, dtype: DTypeLike = None) -> np.ndarray:
+        if dtype is None:
+            dtype = self.dtype
+        array = self.array
+        if isinstance(array, pd.PeriodIndex):
+            with suppress(AttributeError):
+                # this might not be public API
+                array = array.astype("object")
+        return np.asarray(array.values, dtype=dtype)
+
+    @property
+    def shape(self) -> Tuple[int]:
+        return (len(self.array),)
```

Comment:
These len() operations adds up and `shape` is a very popular property in xarray code. I think you should consider caching the property with for example `@pandas.util.cache_readonly` to speed it up quite a bit. 

The cache would have to be cleared if self.array changed size after init, but should that even be allowed?

Reply:
That would probably be welcome indeed. Let's save this for later, we'll need to see if/how we can formalize an (optional) nd-array interface for any `xarray.Index` (which would be required to reuse the index data as coordinate data like it's the case for pandas indexes but maybe other xarray indexes in the future).

> The cache would have to be cleared if self.array changed size after init, but should that even be allowed?

That should not be allowed.